dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/BackwardsReachingDefinitionsAnalysis.java

```diff
@@ -0,0 +1,384 @@
+                }
+                final boolean change = merge(trueStates.computeIfAbsent(jump, k -> new State()), currentState);
+                if (change && (latest == null || latest.getGlobalId() < jump.getGlobalId())) {
+                    latest = jump;
+                }
+            }
+            // If exists, repeat from most distant back jump.
+            // Occurs linear-many times due to guaranteed progress.
+            if (latest != null && label.getGlobalId() < latest.getGlobalId()) {
+                return latest;
+            }
+        }
+        if (event instanceof CondJump jump && !jump.isDead()) {
+            if (jump.isGoto()) {
+                currentState.clear();
+            } else {
+                merge(falseStates.computeIfAbsent(jump, k -> new State()), currentState);
+            }
+            merge(currentState, trueStates.getOrDefault(jump, new State()));
+        }
+        if (event instanceof RegWriter writer) {
+            final Register register = writer.getResultRegister();
+            final StateInfo state = writer.cfImpliesExec() ? currentState.remove(register) : currentState.get(register);
+            if (state != null) {
+                writerMap.get(writer).readers.addAll(state.mayReaders);
+            }
+        }
+        if (event instanceof RegReader reader) {
+            update(currentState, reader);
+        }
+        return null;
+    }
+
+    private static void update(State states, RegReader reader) {
+        for (Register register : reader.getRegisterReads().stream().map(Register.Read::register).distinct().toList()) {
+            final StateInfo state = states.computeIfAbsent(register, k -> new StateInfo());
+            state.mayReaders = state.copyOnWrite ? new HashSet<>(state.mayReaders) : state.mayReaders;
+            state.copyOnWrite = false;
+            state.mayReaders.add(reader);
+        }
+    }
+
+    private static boolean merge(State target, State source) {
+        boolean change = false;
+        for (Map.Entry<Register, StateInfo> entry : source.entrySet()) {
+            change |= merge(target.computeIfAbsent(entry.getKey(), k -> new StateInfo()), entry.getValue());
+        }
+        return change;
+    }
+
+    private static boolean merge(StateInfo target, StateInfo source) {
+        if (target.mayReaders.containsAll(source.mayReaders)) {
+            return false;
+        }
+        assert !source.mayReaders.isEmpty();
+        if (target.mayReaders.isEmpty()) {
+            target.mayReaders = source.mayReaders;
+            target.copyOnWrite = source.copyOnWrite = true;
+            return true;
+        }
+        target.mayReaders = target.copyOnWrite ? new HashSet<>(target.mayReaders) : target.mayReaders;
+        target.copyOnWrite = false;
+        target.mayReaders.addAll(source.mayReaders);
+        return true;
+    }
+
+    private void analyzeMust(ExecutionAnalysis exec) {
+        // Require that function is unrolled.
```

Comment:
Can this be programmatically checked? Maybe in the caller?

Reply:
I would assume this is free because `ExecutionAnalysis` already assumes this. In fact, at this point in time we are only working over threads.

Reply:
Yes, it is.  But to keep safe, I left this comment.